doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+- `commit_transaction`:
+    1. Use the write handle to compare and append the batch of changes into persist.
+    2. Update the in memory upper to the new upper.
+    3. Update in-memory object state.
+- `set_X`:
+    1. For each object create the following batch of changes:
+        - If the key of the object already exists in memory, add a retraction to the batch.
+        - Add an addition to the batch of the new value.
+    2. Use the write handle to compare and append the batch of changes into persist.
+    3. Update the in memory upper to the new upper.
+    4. Update in-memory object state.
+
+#### Misc
+
+- `confirm_leadership`: Check that the persist shard upper is equal to the upper cached in memory.
+- `dump`: Convert the in-memory state to a JSON string.
+
+### Catalog Transactions
+
+The current
+[catalog transactions](https://github.com/MaterializeInc/materialize/blob/b5aa928960aa68c031caa8ff3bfab0ff53220361/src/adapter/src/catalog/storage.rs#L892-L916)
+will remain mostly the same. It works by reading the entire durable catalog state into memory,
+making changes to the state in memory, then committing all changes at once to the backing durable
+storage. The few changes needed are:
+
+- It will be the responsibility of the `DurableCatalogState` to commit the `Transaction` instead of
+  the `Transaction` committing itself.
+- `Transaction` will not hold onto a mutable reference to a `Stash`. Though it will need a mutable
+  reference to some part of `DurableCatalogState` to prevent multiple concurrent transactions.
+- To avoid cloning the entire catalog state in the persist implementation, we'll
+  update `TableTransaction`s to use `CoW`s.
+
+### Catalog Debug Tool
+
+The catalog debug tool is modeled after the stash debug tool and allows users to inspect and modify
+the durable catalog state. It will provide the following functionality (descriptions are mostly
+taken from the existing stash debug tool):
+
+- Dump: Dumps the catalog contents in a human-readable format. Includes JSON for each key and value
+  that can be hand edited and then passed to the `edit` or `delete` commands.
+    - [optional arg] `target`: Write output to specified path. Default to stdout.
+- Edit: Edits a single catalog object.
+    - [arg] `object-type`: name of the object type to edit.
+    - [arg] `key`: JSON encoded key of object.
+    - [arg] `value`: JSON encoded value of object.
+- Delete: Deletes a single catalog object.
+    - [arg] `object-type`: name of the object type to edit.
+    - [arg] `key`: JSON encoded key of object.
+- Upgrade Check: Checks if the specified catalog could be upgraded from its state to the catalog at
+  the version of this binary. Prints a success message or error message. Exits with 0 if the upgrade
+  would succeed, otherwise non-zero. Can be used on a running `environmentd`. Operates without
+  interfering with it or committing any data to that catalog.
```

Comment:
For the persist impl, do read leases ever cause other accessors to be delayed? The current stash-debug had to learn about [low priority transaction](https://github.com/MaterializeInc/materialize/blob/9aa1febad6136091f789bc7c8f4d10f79eaedaef/src/stash/src/postgres.rs#L613) in order to prevent retry or blocking errors from a running envd.

Reply:
Going to outsource this to @danhhz again. 

Intuitively I would think no, a read at timestamp `ts1` won't block another read at timestamp `ts2`.